drivers/gpu/drm/tests/drm_framebuffer_test.c

```diff
@@ -432,8 +433,133 @@ static void drm_test_framebuffer_modifiers_not_supported(struct kunit *test)
 	KUNIT_EXPECT_EQ(test, 0, buffer_created);
 }
 
+/* Parameters for testing drm_framebuffer_check_src_coords function */
+struct check_src_coords_case {
+	const char *name; /* Description of the parameter case */
+	const int expect; /* Expected returned value by the function */
+
+	/* All function args */
+	const uint32_t src_x;
+	const uint32_t src_y;
+	const uint32_t src_w;
+	const uint32_t src_h;
+	const struct drm_framebuffer fb;
+};
+
+static const struct check_src_coords_case check_src_coords_cases[] = {
+	/* Regular case where the source just fit in the framebuffer */
+	{ .name = "source inside framebuffer with normal sizes and coordinates",
+	  .expect = 0,
+	  .src_x = 500 << 16, .src_y = 700 << 16,
+	  .src_w = 100 << 16, .src_h = 100 << 16,
+	  .fb = { .width = 600, .height = 800 }
+	},
+	{ .name = "out of bound by both x and y with normal sizes and coordinates",
+	  .expect = -ENOSPC,
+	  .src_x = (500 << 16) + 1, .src_y = (700 << 16) + 1,
+	  .src_w = 100 << 16, .src_h = 100 << 16,
+	  .fb = { .width = 600, .height = 800 }
+	},
+	/* From here, cases involving only x axis */
+	{ .name = "out of bound by x with normal width and x",
```

Comment:
this seems a bit contradictory, is x out of bound or normal?

Reply:
The word "normal" here means that the combination of src_x and src_w is both within the allowed individual values and that the test fail due to its combination. I guess there should be a better description for it, I just couldn't realize it.

Reply:
If you presented to me:

1. Success: fb.width = 600, src_x=0, src_w=600
2. Fail: fb.width = 600, src_x=0, src_w=601
3. Fail: fb.width = 600, src_x=1, src_w=600

It would be clearer (for me).

And I would read as

1. Success: src and fb fit
2. Fail: overflowing fb with source x-axis width 
3. Fail: overflowing fb with source x-axis source

Reply:
Extending this, I would present

800x600 because that is an usual widescreen size

1. fb.width = 800, fb.height = 600, src_x=0, src_w=800, src_y=0, src_h=600
2. fb.width = 800, fb.height = 600, src_x=1, src_w=800, src_y=0, src_h=600
3. fb.width = 800, fb.height = 600, src_x=0, src_w=801, src_y=0, src_h=600
4. fb.width = 800, fb.height = 600, src_x=0, src_w=800, src_y=1, src_h=600
5. fb.width = 800, fb.height = 600, src_x=0, src_w=800, src_y=0, src_h=601

Too improve even further, I would actually write the deltas (0, +1) as params, so that it would avoid the whole noise from `<< 16`

Reply:
If you want to show a success with src != 0
1.1 fb.width = 800, fb.height = 600, src_x=1, src_w=800-1, src_y=0, src_h=600
1.2 fb.width = 800, fb.height = 600, src_x=0, src_w=800, src_y=1, src_h=600-1

Reply:
> Too improve even further, I would actually write the deltas (0, +1) as params, so that it would avoid the whole noise from `<< 16`

I'm not sure if I understood correctly, but what you're proposing is to have new members on the param struct, like dsrc_w, dsrc_h, dsrc_x and dsrc_y to represent the deltas or it''s to just write everything without `<< 16` and let the test function do that?

In the second case, the delta will be shifted too, for example: `src_x = 100 + 1` will actually result on `src_x = (100 + 1) << 16`, which may not cover the overflowing "by one". If something goes wrong on the tested function so that it reports that a source fits inside the framebuffer when it actually overflows on the subpixel range (reporting that `src_x = (100 << 16) + 1; src_w = (500 << 16)` fit on `fb.width = 600`, for example), the test written that way will not catch it.

So, I assume that the first solution I mentioned would be ok, but isn't that adding some noise too?

Reply:
I guess I was thinking of keeping the values 600<<16 and 800<<16 static in the test and use only deltas in the params

Reply:
Ahh, I think it could be nice, thanks for the suggestion.